RecoParticleFlow/PFClusterProducer/plugins/LegacyPFClusterProducer.cc

```diff
@@ -0,0 +1,251 @@
+      }
+      {
+        edm::ParameterSetDescription bar;
+        bar.add<double>("corrTermLowE", 0.);
+        bar.add<double>("threshLowE", 6.);
+        bar.add<double>("noiseTerm", 21.86);
+        bar.add<double>("constantTermLowE", 4.24);
+        bar.add<double>("noiseTermLowE", 8.);
+        bar.add<double>("threshHighE", 15.);
+        bar.add<double>("constantTerm", 2.82);
+        pfClusterBuilder.add("timeResolutionCalcBarrel", bar);
+      }
+      {
+        edm::ParameterSetDescription bar;
+        bar.add<double>("corrTermLowE", 0.);
+        bar.add<double>("threshLowE", 6.);
+        bar.add<double>("noiseTerm", 21.86);
+        bar.add<double>("constantTermLowE", 4.24);
+        bar.add<double>("noiseTermLowE", 8.);
+        bar.add<double>("threshHighE", 15.);
+        bar.add<double>("constantTerm", 2.82);
+        pfClusterBuilder.add("timeResolutionCalcEndcap", bar);
+      }
+      {
+        edm::ParameterSetDescription bar;
+        pfClusterBuilder.add("positionReCalc", bar);
+      }
+      {
+        edm::ParameterSetDescription bar;
+        pfClusterBuilder.add("energyCorrector", bar);
+      }
+      desc.add("pfClusterBuilder", pfClusterBuilder);
+    }
+    descriptions.addWithDefaultLabel(desc);
+  }
+
+  // the actual algorithm
+  std::unique_ptr<PFCPositionCalculatorBase> _positionCalc;
+  std::unique_ptr<PFCPositionCalculatorBase> _allCellsPositionCalc;
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  const edm::EDGetTokenT<reco::PFClusterHostCollection> alpakaPfClustersToken;
+  const edm::EDGetTokenT<reco::PFRecHitFractionHostCollection> alpakaPfRHFractionToken;
+  const edm::EDGetTokenT<reco::PFRecHitHostCollection> InputPFRecHitSoA_Token_;
+  const edm::ESGetToken<reco::PFClusterParamsHostCollection, JobConfigurationGPURecord> pfClusParamsToken;
+  const edm::EDPutTokenT<reco::PFClusterCollection> legacyPfClustersToken;
+  const edm::EDGetTokenT<reco::PFRecHitCollection> recHitsLabel;
+
+  int nRH = 0;
+};
+
+void LegacyPFClusterProducer::produce(edm::Event& event, const edm::EventSetup& setup) {
+  const reco::PFRecHitHostCollection& pfRecHits = event.get(InputPFRecHitSoA_Token_);
+
+  reco::PFClusterHostCollection const& clusterSoA = event.get(alpakaPfClustersToken);
+  reco::PFRecHitFractionHostCollection const& fractionSoA = event.get(alpakaPfRHFractionToken);
+
+  auto const& alpakaPfClusters = clusterSoA.const_view();
+  auto const& alpakaPfRhFrac = fractionSoA.const_view();
+
+  nRH = pfRecHits.view().size();
+  reco::PFClusterCollection out;
+  out.reserve(nRH);
+
+  auto const rechitsHandle = event.getHandle(recHitsLabel);
+
+  std::vector<int> seedlist1;
+  std::vector<int> seedlist2;
+  seedlist1.reserve(alpakaPfClusters.nSeeds());
+  for (int i = 0; i < alpakaPfClusters.nSeeds(); i++) {
+    seedlist1.push_back(alpakaPfClusters[i].seedRHIdx());
+  }
+
+  // Build PFClusters in legacy format
+  std::unordered_map<int, int> nTopoSeeds;
```

Comment:
Why an `std::unordered_map` instead of a `vector` ?
How sparse is the `topoId()` ?
What is the largest value it can have ?

Reply:
I think this was done for simplicity of the frequency counting? The topoId values range from 0 to usually 400 depending on the event. The `topoId()` SoA member only contains a subset of the whole range, for example in one QCD sample, there are 75 which become clusters.

Reply:
OK, if you know the size a priory, I'd suggest changing this to
```suggestion
  std::vector<int> nTopoSeeds(size, 0);
```

Reply:
Well the topoIds which become clusters are not exactly continuous, and could be something like `[0, 0, 25, 26, 68, 68, 68, 101, ...]` and our unordered map would only have as many elements as unique IDs. I think switching to a vector here we would need some modification to the cluster formation and to then loop over the topoIds again to count how many seeds are in each. At least in our use case we believe the unordered map to be a bit more well suited and simplifies the code structure.

Reply:
I think I understand.

As an example, let's assume 75 different `topoId`, versus a maximum of 400 potential `topoId`.

An `unordered_map` with 75 `int` should use between 2100 and 2700 bytes (see https://stackoverflow.com/a/25438497/2050986 for a discussion), and accessing an element requires evaluating a hash function and 3 memory indirections.
This leaves out the time to fill the `unoredered_map`.

A `vector` with 400 `int` uses 1624 bytes, and accessing an element requires a single memory indirection.

So, a `vector` should use less memory and be more efficient :-)

Reply:
Checking on a TTBar sample we have topoIds up to ~2500, and ~400 becoming clusters. The max value of topoId is not known at compile time and would need to be calculated, but we can probably do so in the kernel. For these larger cases that arise, is a vector still the recommendation?

Reply:
@jsamudio If ~400 refers to the number of PF clusters, the number of topological clusters (which will be the number of unordered_map elements) is much smaller than that?

Reply:
> Checking on a TTBar sample we have topoIds up to ~2500, and ~400 becoming clusters.

In that case, what is the value of `pfClusterSoA.nSeeds()` ?


> For these larger cases that arise, is a vector still the recommendation?

I've spent some time reading the implementation of `std::unordered_map` in GCC 12.
An ideally "balanced" map (where the number of buckets is the same as the number of elements) uses
```
24 + size * (16 + sizeof(key) + sizeof(value))  bytes
```

A non-balanced map could have more buckets than needed (and so waste some memory) or less (and so save a little bit of memory, but be slower to fill and access).

An `std::vector` will always use
```
24 + capacity * sizeof(value)  bytes
```

In this case both `key` and `value` are `int`, so 4 bytes each.
So the `std::unordered_map<int,int>` takes `24 + size * 24 bytes`, while the `std::vector<int>` takes `24 + capacity * 4 bytes`.

So the `std::vector` will use less memory as long as it is filled for 1/6th or more.
And it will always be faster to read, and faster to write if the capacity is reserved up front.

If you don't know the capacity but can make a good guess (say, within a factor 2), that's good enough - it will either reallocate once, or waste some memory, but still be better off than the `std::unoredered_map`.

Considering also the speed, I think the `std::unoredered_map` will be a better choice only if it is very sparse, with 1/10th of the keys being filled or less.

Reply:
If you prefer to do a minimal memory allocation, you can also do a first pass to find the maximum `topoId`:
```c++
int maxTopoId = -1;
for (int i = 0; i < pfClusterSoA.nSeeds(); i++) {
  maxTopoId = std::max(maxTopoId, pfClusterSoA[i].topoId());
}
std::vector<int> nTopoSeeds(maxTopoId + 1, 0);
```

I'm not sure if that is going to be fast than a simple guess based on `pfClusterSoA.nSeeds()`, though.